core/cont/inc/TCollection.h

```diff
@@ -296,78 +298,20 @@ namespace Internal {
-
-   /// This is a workaround against ClassDefInline not supporting classes
-   /// missing their default constructor or having them private.
-   template <class T>
-   friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
-
-   TRangeDynCastIterator() = default;
-
-public:
-   using TIter::TIter;
-   TRangeDynCastIterator(const TIter &iter) : TIter(iter) {}
-
-   Containee *operator()() = delete;
-
-   Containee *Next() { return dynamic_cast<Containee *>(TIter::Next()); }
-   Containee *operator*() const { return dynamic_cast<Containee *>(TIter::operator*()); }
-
-   ClassDefInline(TRangeDynCastIterator, 0);
-};
-
 } // namespace Internal
 
 namespace Detail {
 
-/// @brief TTypedIter is a typed version of TIter.
-///
-/// This requires the collection to contains elements of the type requested
-/// (or a derived class).  Any deviation from this expectation
-/// will only be caught/reported by an assert in debug builds.
-///
-/// This is best used with a TClonesArray, for other cases prefered TRangeDynCast.
-///
-/// The typical use is:
-/// ```{.cpp}
-///    TTypedIter<TBaseClass> next(tbaseClassClonesArrayPtr);
-///    while(auto bcl = next()) {
-///       ... use bcl as a TBaseClass*
-///    }
-/// ```
-template <class Containee> // Containee must derive from TObject.
-class TTypedIter : public TIter {
-   static_assert(std::is_base_of<TObject, Containee>::value, "Containee type must inherit from TObject");
-
-   /// This is a workaround against ClassDefInline not supporting classes
-   /// missing their default constructor or having them private.
-   template <class T>
-   friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
-
-   TTypedIter() = default;
-
-   static Containee *StaticCast(TObject *obj)
-   {
-      assert(!obj || ROOT::Internal::ContaineeInheritsFrom(obj->IsA(), Containee::Class()));
-      return static_cast<Containee *>(obj);
-   }
-
+template <typename T, bool isDynamic>
+class TRangeCast : public ROOT::RRangeCast<T*, isDynamic, TCollection const&> {
 public:
-   using TIter::TIter;
-   TTypedIter(const TIter &iter) : TIter(iter) {}
-
-   Containee *operator()() { return StaticCast(TIter::Next()); }
-   Containee *Next() { return StaticCast(TIter::Next()); }
-   Containee *operator*() const { return StaticCast(TIter::operator*()); }
-
-   ClassDefInline(TTypedIter, 0);
+   TRangeCast(TCollection const& col) : ROOT::RRangeCast<T*, isDynamic, TCollection const&>{col} {
+      static_assert(std::is_base_of<TObject, T>::value, "Containee type must inherit from TObject");
+   }
+   TRangeCast(TCollection const* col) : TRangeCast{col != nullptr ? *col : ROOT::Internal::EmptyCollection()} {}
 };
 
-/// @brief TRangeStaticCast is an adaptater class that allows the typed iteration
+/// @brief TRangeStaticCast is an adapter class that allows the typed iteration
 /// through a TCollection.  This requires the collection to contains element
```

Comment:
`contains --> contain`

Consider adding the
```c++
/// \tparam T ...
```

Reply:
Is it only for `TCollection`? I don't think so. Also, maybe update the code example below to not use the  asserts?

Reply:
True, the asserts are done in `TRangeStaticCast` itself! And indeed `TRangeStaticCast` is only for `TCollections`.